pkg/router/metrics/haproxy/haproxy.go

```diff
@@ -0,0 +1,648 @@
+			if namespace, name, ok := parseNameSegment(value); ok {
+				e.exportCsvFields(e.serverMetrics, csvRow, server, namespace, name, pod, service)
+				return
+			}
+		}
+		e.exportCsvFields(e.serverMetrics, csvRow, server, "", "", pod, service)
+	}
+}
+
+// knownServerSegment takes a server name that has a known prefix and returns
+// the pod, service, and simpler service name label for that type. If the prefix does not
+// match false is returned.
+func knownServerSegment(value string) (string, string, string, bool) {
+	if i := strings.Index(value, ":"); i != -1 {
+		switch value[:i] {
+		case "ept":
+			if service, server, ok := parseNameSegment(value[i+1:]); ok {
+				return "", service, server, true
+			}
+		case "pod":
+			if pod, remainder, ok := parseNameSegment(value[i+1:]); ok {
+				if service, server, ok := parseNameSegment(remainder); ok {
+					return pod, service, server, true
+				}
+			}
+		}
+	}
+	return "", "", value, false
+}
+
+// knownBackendSegment takes a backend name that has a known prefix and returns
+// the preferred metric label for that type and the remainder. If the prefix does not
+// match false is returned.
+func knownBackendSegment(value string) (string, string, bool) {
+	if i := strings.Index(value, ":"); i != -1 {
+		switch value[:i] {
+		case "be_http":
+			return "http", value[i+1:], true
+		case "be_edge_http":
+			return "https-edge", value[i+1:], true
+		case "be_secure":
+			return "https", value[i+1:], true
+		case "be_tcp":
+			return "tcp", value[i+1:], true
+		}
+	}
+	return "", "", false
+}
+
+// parseNameSegment splits a name by the first colon and returns both segments. If
+// no colon is found it returns false.
+func parseNameSegment(value string) (string, string, bool) {
+	i := strings.Index(value, ":")
+	if i == -1 {
+		return "", "", false
+	}
+	return value[:i], value[i+1:], true
+}
+
+func parseStatusField(value string) int64 {
+	switch value {
+	case "UP", "UP 1/3", "UP 2/3", "OPEN", "no check":
+		return 1
+	case "DOWN", "DOWN 1/2", "NOLB", "MAINT":
+		return 0
+	}
+	return 0
+}
+
+func (e *Exporter) exportCsvFields(metrics metrics, csvRow []string, labels ...string) {
+	for fieldIdx, metric := range metrics {
+		valueStr := csvRow[fieldIdx]
+		if valueStr == "" {
+			continue
+		}
+
+		var value int64
+		switch fieldIdx {
+		case statusField:
+			value = parseStatusField(valueStr)
+		default:
+			var err error
+			value, err = strconv.ParseInt(valueStr, 10, 64)
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("can't parse CSV field value %s: %v", valueStr, err))
+				e.csvParseFailures.Inc()
+				continue
+			}
+		}
+		metric.WithLabelValues(labels...).Set(float64(value))
+	}
+}
+
+// filterMetrics returns the set of server metrics specified by the export array.
+func filterMetrics(export []int, available metrics) metrics {
+	metrics := map[int]*prometheus.GaugeVec{}
+	if len(export) == 0 {
+		return metrics
+	}
+
+	selected := map[int]struct{}{}
```

Comment:
We do - but it's not necessarily any easier to read.  I'm a stickler in core code, but I'm not sure the extra dependencies will make sense in all cases.